### PR TITLE
PMM-6851 add edit cluster parameters modal

### PR DESCRIPTION
**Pull request by lunaticusgreen**
[PMM-6851](https://jira.percona.com/browse/PMM-6851)

### CODE REVIEW

File: pmm-app/src/pmm-dbaas/components/DBCluster/DBCluster.hooks.ts [hunk 1]
```diff
@@ -68,13 +65,8 @@ const getClusters = async (kubernetes: Kubernetes[], databaseType: Databases): P
...
 
-    const resultClusters = clusters.map(
-      (cluster) => dbClusterService.toModel(
-        cluster,
-        kubernetes[index].kubernetesClusterName,
-        databaseType,
-      ),
-    );
+    // eslint-disable-next-line max-len
+    const resultClusters = clusters.map((cluster) => dbClusterService.toModel(cluster, kubernetes[index].kubernetesClusterName, databaseType));
```
Comment (tiagomotasantos) [inline pmm-app/src/pmm-dbaas/components/DBCluster/DBCluster.hooks.ts:69]: Why this change? Was it the IDE? I think it's better like it was to avoid this eslint comments. What do you think?

File: pmm-app/src/pmm-dbaas/components/DBCluster/EditDBClusterModal/DBClusterAdvancedOptions/DBClusterAdvancedOptions.constants.ts [hunk 2]
```diff
@@ -0,0 +1,38 @@
+import { SelectableValue } from '@grafana/data';
```
Comment (tiagomotasantos) [inline pmm-app/src/pmm-dbaas/components/DBCluster/EditDBClusterModal/DBClusterAdvancedOptions/DBClusterAdvancedOptions.constants.ts:1]: This file doesn't need to be duplicated. Use the one in the AddDBClusterModal folder instead.

File: pmm-app/src/pmm-dbaas/components/DBCluster/EditDBClusterModal/DBClusterAdvancedOptions/DBClusterAdvancedOptions.styles.ts [hunk 3]
```diff
@@ -0,0 +1,29 @@
+import { stylesFactory } from '@grafana/ui';
```
Comment (tiagomotasantos) [inline pmm-app/src/pmm-dbaas/components/DBCluster/EditDBClusterModal/DBClusterAdvancedOptions/DBClusterAdvancedOptions.styles.ts:1]: This file doesn't need to be duplicated. Use the one in the AddDBClusterModal folder instead. (Unless there are specific styles for the edit)

File: pmm-app/src/pmm-dbaas/components/DBCluster/EditDBClusterModal/DBClusterAdvancedOptions/DBClusterAdvancedOptions.types.ts [hunk 4]
```diff
@@ -0,0 +1,11 @@
+export enum DBClusterTopology {
```
Comment (tiagomotasantos) [inline pmm-app/src/pmm-dbaas/components/DBCluster/EditDBClusterModal/DBClusterAdvancedOptions/DBClusterAdvancedOptions.types.ts:1]: This file doesn't need to be duplicated. Use the one in the AddDBClusterModal folder instead.

File: pmm-app/src/pmm-dbaas/components/DBCluster/EditDBClusterModal/EditDBClusterModal.tsx [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+      setVisible(false);
+
+      return;
+    }
+
+    try {
+      const dbClusterService = DBClusterServiceFactory.newDBClusterService(databaseType.value);
+
+      await dbClusterService.updateDBCluster({
+        databaseType,
```
Comment (tiagomotasantos) [inline pmm-app/src/pmm-dbaas/components/DBCluster/EditDBClusterModal/EditDBClusterModal.tsx:35]:
```suggestion
        databaseType: databaseType.value,
```

I know we are not using it but let's make it right. It might be needed in the future.

File: pmm-app/src/pmm-dbaas/components/DBCluster/EditDBClusterModal/EditDBClusterModal.tsx [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+    clusterParameters.topology = clusterParameters.clusterSize > 1 ? DBClusterTopology.cluster : DBClusterTopology.single;
+    clusterParameters[EditDBClusterFields.nodes] = clusterParameters.clusterSize;
+    clusterParameters[EditDBClusterFields.single] = clusterParameters.clusterSize;
+
+    const isMatchSize = (type) => DEFAULT_SIZES[type].cpu === clusterParameters.cpu
+      && DEFAULT_SIZES[type].memory === clusterParameters.memory
+      && DEFAULT_SIZES[type].disk === clusterParameters.disk;
+
+    if (isMatchSize('small')) {
+      clusterParameters.resources = 'small';
```
Comment (tiagomotasantos) [inline pmm-app/src/pmm-dbaas/components/DBCluster/EditDBClusterModal/EditDBClusterModal.tsx:74]: There is an enum for this values

File: pmm-app/src/pmm-dbaas/components/DBCluster/EditDBClusterModal/EditDBClusterModal.tsx [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+  const [initialValues, setInitialValues] = useState({});
+
+  useEffect(() => {
+    if (!selectedCluster) {
+      setVisible(false);
+
+      return;
+    }
+
+    const clusterParameters: any = { ...selectedCluster };
```
Comment (tiagomotasantos) [inline pmm-app/src/pmm-dbaas/components/DBCluster/EditDBClusterModal/EditDBClusterModal.tsx:59]: Try FormRenderProps here instead of any. Might be more specific.

File: pmm-app/src/pmm-dbaas/components/DBCluster/EditDBClusterModal/EditDBClusterModal.tsx [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+    } else {
+      clusterParameters.resources = 'custom';
+    }
+
+    setInitialValues(clusterParameters);
+  }, [selectedCluster]);
+
+  return (
+    // eslint-disable-next-line max-len
+    <Modal title={`Edit cluster ${selectedCluster?.clusterName} (${selectedCluster?.databaseType})`} isVisible={isVisible} onClose={() => setVisible(false)}>
```
Comment (tiagomotasantos) [inline pmm-app/src/pmm-dbaas/components/DBCluster/EditDBClusterModal/EditDBClusterModal.tsx:88]: Break to different lines and remove the comment to be consistent.
nit: put a function in Messages file to construct this string

File: pmm-app/src/pmm-dbaas/components/DBCluster/EditDBClusterModal/EditDBClusterModal.tsx [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+      <FormFinal
+        onSubmit={onSubmit}
+        initialValues={initialValues}
+        validate={() => undefined}
```
Comment (tiagomotasantos) [inline pmm-app/src/pmm-dbaas/components/DBCluster/EditDBClusterModal/EditDBClusterModal.tsx:92]: Does it really need this validate here?

File: pmm-app/src/pmm-dbaas/components/DBCluster/EditDBClusterModal/EditDBClusterModal.tsx [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+        render={(renderProps) => (
+          <form onSubmit={renderProps.handleSubmit} className="discovery-instance-form app-theme-dark">
```
Comment (tiagomotasantos) [inline pmm-app/src/pmm-dbaas/components/DBCluster/EditDBClusterModal/EditDBClusterModal.tsx:94]: I think this classes can be removed.

File: pmm-app/src/pmm-dbaas/components/DBCluster/EditDBClusterModal/EditDBClusterModal.types.ts [hunk 11]
```diff
@@ -0,0 +1,21 @@
+import { DBCluster } from '../DBCluster.types';
```
Comment (tiagomotasantos) [inline pmm-app/src/pmm-dbaas/components/DBCluster/EditDBClusterModal/EditDBClusterModal.types.ts:1]: This file doesn't need to be duplicated. Use the one in the AddDBClusterModal folder instead.

File: pmm-app/src/pmm-dbaas/components/DBCluster/PSMDB.service.ts [hunk 12]
```diff
@@ -21,7 +21,7 @@ const DBCLUSTER_STATUS_MAP = {
 };
 
 export class PSMDBService extends DBClusterService {
-  getDBClusters(kubernetes: Kubernetes): Promise<DBClusterPayload> {
+  async getDBClusters(kubernetes: Kubernetes): Promise<DBClusterPayload> {
```
Comment (tiagomotasantos) [inline pmm-app/src/pmm-dbaas/components/DBCluster/PSMDB.service.ts:24]: If really needs async here change in the parent too.

File: pmm-app/src/pmm-dbaas/components/DBCluster/PSMDB.service.ts [hunk 13]
```diff
@@ -32,6 +32,29 @@ export class PSMDBService extends DBClusterService {
     );
   }
 
+  updateDBCluster(dbCluster: DBCluster): Promise<void | DBClusterPayload> {
+    const toAPI = (cluster: DBCluster) => ({
```
Comment (tiagomotasantos) [inline pmm-app/src/pmm-dbaas/components/DBCluster/PSMDB.service.ts:36]: Don't forget to remove this toAPI methods like you did in other branch.

File: pmm-app/src/pmm-dbaas/components/DBCluster/XtraDB.service.ts [hunk 14]
```diff
@@ -32,6 +32,29 @@ export class XtraDBService extends DBClusterService {
     );
   }
 
+  updateDBCluster(dbCluster: DBCluster): Promise<void | DBClusterPayload> {
+    const toAPI = (cluster: DBCluster) => ({
```
Comment (tiagomotasantos) [inline pmm-app/src/pmm-dbaas/components/DBCluster/XtraDB.service.ts:36]: Don't forget to remove this :D

File: pmm-app/src/pmm-dbaas/components/Kubernetes/Kubernetes.service.ts [hunk 15]
```diff
@@ -8,6 +8,14 @@ import {
 
 export const KubernetesService = {
   getKubernetes() {
+    // return {
```
Comment (tiagomotasantos) [inline pmm-app/src/pmm-dbaas/components/Kubernetes/Kubernetes.service.ts:11]: Comment :D ♟

File: pmm-app/src/pmm-dbaas/components/DBCluster/EditDBClusterModal/EditDBClusterModal.tsx [hunk 16]
```diff
@@ -0,0 +1,103 @@
...
+      onDBClusterChanged();
+    } catch (e) {
+      console.error(e);
+    }
+  };
+
+  const [initialValues, setInitialValues] = useState({});
+
+  useEffect(() => {
+    if (!selectedCluster) {
```
Comment (tiagomotasantos) [inline pmm-app/src/pmm-dbaas/components/DBCluster/EditDBClusterModal/EditDBClusterModal.tsx:53]: Does this check need to be repeated twice in this component?

File: pmm-app/src/pmm-dbaas/components/DBCluster/EditDBClusterModal/DBClusterAdvancedOptions/DBClusterAdvancedOptions.tsx [hunk 17]
```diff
@@ -0,0 +1,137 @@
...
+      <HorizontalGroup justify="center" spacing="md">
+        <LoaderButton
+          data-qa="dbcluster-update-cluster-button"
+          size="md"
+          variant="primary"
+          disabled={!valid || pristine || submitting}
+          loading={submitting}
+          className={styles.createButton}
+        >
+          Save changes
```
Comment (tiagomotasantos) [inline pmm-app/src/pmm-dbaas/components/DBCluster/EditDBClusterModal/DBClusterAdvancedOptions/DBClusterAdvancedOptions.tsx:132]: Messages file

File: pmm-app/src/pmm-dbaas/components/DBCluster/EditDBClusterModal/DBClusterAdvancedOptions/DBClusterAdvancedOptions.tsx [hunk 18]
```diff
@@ -0,0 +1,137 @@
+import React, {
```
Comment (tiagomotasantos) [inline pmm-app/src/pmm-dbaas/components/DBCluster/EditDBClusterModal/DBClusterAdvancedOptions/DBClusterAdvancedOptions.tsx:1]: If there are no changes from the other DBClusterAdvancedOptions component there's no need to duplicate. Just remove this and the tests too will simplify this a lot.

**Review by tiagomotasantos — CHANGES_REQUESTED** [review]

File: pmm-app/src/pmm-dbaas/components/DBCluster/EditDBClusterModal/DBClusterAdvancedOptions/DBClusterAdvancedOptions.constants.ts (hunk 2 above)
Comment (lunaticusgreen) [reply pmm-app/src/pmm-dbaas/components/DBCluster/EditDBClusterModal/DBClusterAdvancedOptions/DBClusterAdvancedOptions.constants.ts:1]: It's a philosophical question, I would say that in this case, it's better to violate DRY than create a too tight coupling between AddDBClusterModal and EditDBClusterModal. These components a little different, and maybe the difference will be bigger in the future so it's dangerous to use dependencies directly. What do you think?

Comment (tiagomotasantos) [reply pmm-app/src/pmm-dbaas/components/DBCluster/EditDBClusterModal/DBClusterAdvancedOptions/DBClusterAdvancedOptions.constants.ts:1]: Totally agree with you. That's why we should keep the duplication in things like component itself. But constants and types? Do we really need the duplication? I doubt that we will have different topologies and resources in both modals for example. It's more a subjective question rather than philosophical 😀 So yeah it can go both sides, just pick one. Either one is fine to me, was just raising this concern.

Comment (lunaticusgreen) [reply pmm-app/src/pmm-dbaas/components/DBCluster/EditDBClusterModal/DBClusterAdvancedOptions/DBClusterAdvancedOptions.constants.ts:1]: Decided to discuss it on a next sync call

File: pmm-app/src/pmm-dbaas/components/DBCluster/EditDBClusterModal/DBClusterAdvancedOptions/DBClusterAdvancedOptions.styles.ts (hunk 3 above)
Comment (lunaticusgreen) [reply pmm-app/src/pmm-dbaas/components/DBCluster/EditDBClusterModal/DBClusterAdvancedOptions/DBClusterAdvancedOptions.styles.ts:1]: Decided to discuss it on a next sync call

File: pmm-app/src/pmm-dbaas/components/DBCluster/EditDBClusterModal/EditDBClusterModal.types.ts (hunk 11 above)
Comment (lunaticusgreen) [reply pmm-app/src/pmm-dbaas/components/DBCluster/EditDBClusterModal/EditDBClusterModal.types.ts:1]: Decided to discuss it on a next sync call

File: pmm-app/src/pmm-dbaas/components/DBCluster/PSMDB.service.ts (hunk 13 above)
Comment (lunaticusgreen) [reply pmm-app/src/pmm-dbaas/components/DBCluster/PSMDB.service.ts]: done

File: pmm-app/src/pmm-dbaas/components/DBCluster/XtraDB.service.ts (hunk 14 above)
Comment (lunaticusgreen) [reply pmm-app/src/pmm-dbaas/components/DBCluster/XtraDB.service.ts]: done

File: pmm-app/src/pmm-dbaas/components/Kubernetes/Kubernetes.service.ts (hunk 15 above)
Comment (lunaticusgreen) [reply pmm-app/src/pmm-dbaas/components/Kubernetes/Kubernetes.service.ts]: done

File: pmm-app/src/pmm-dbaas/components/DBCluster/EditDBClusterModal/DBClusterAdvancedOptions/DBClusterAdvancedOptions.tsx (hunk 17 above)
Comment (lunaticusgreen) [reply pmm-app/src/pmm-dbaas/components/DBCluster/EditDBClusterModal/DBClusterAdvancedOptions/DBClusterAdvancedOptions.tsx:132]: done

File: pmm-app/src/pmm-dbaas/components/DBCluster/EditDBClusterModal/DBClusterAdvancedOptions/DBClusterAdvancedOptions.tsx (hunk 18 above)
Comment (lunaticusgreen) [reply pmm-app/src/pmm-dbaas/components/DBCluster/EditDBClusterModal/DBClusterAdvancedOptions/DBClusterAdvancedOptions.tsx:1]: Decided to discuss it on a next sync call

File: pmm-app/src/pmm-dbaas/components/DBCluster/PSMDB.service.ts (hunk 12 above)
Comment (lunaticusgreen) [reply pmm-app/src/pmm-dbaas/components/DBCluster/PSMDB.service.ts:24]: no need, it was added for testing purpose, removed

File: pmm-app/src/pmm-dbaas/components/DBCluster/EditDBClusterModal/EditDBClusterModal.tsx (hunk 10 above)
Comment (lunaticusgreen) [reply pmm-app/src/pmm-dbaas/components/DBCluster/EditDBClusterModal/EditDBClusterModal.tsx:94]: done

File: pmm-app/src/pmm-dbaas/components/DBCluster/DBCluster.hooks.ts (hunk 1 above)
Comment (lunaticusgreen) [reply pmm-app/src/pmm-dbaas/components/DBCluster/DBCluster.hooks.ts:67]: fixed, it was autoformatting from eslint

File: pmm-app/src/pmm-dbaas/components/DBCluster/EditDBClusterModal/EditDBClusterModal.tsx (hunk 6 above)
Comment (lunaticusgreen) [reply pmm-app/src/pmm-dbaas/components/DBCluster/EditDBClusterModal/EditDBClusterModal.tsx:74]: wow, thanks, didn't notice it

File: pmm-app/src/pmm-dbaas/components/DBCluster/EditDBClusterModal/EditDBClusterModal.tsx (hunk 8 above)
Comment (lunaticusgreen) [reply pmm-app/src/pmm-dbaas/components/DBCluster/EditDBClusterModal/EditDBClusterModal.tsx]: Fixed, not moved it to the message because I'm not sure if we ever did that and don't want to overload Messages file

File: pmm-app/src/pmm-dbaas/components/DBCluster/EditDBClusterModal/EditDBClusterModal.tsx (hunk 9 above)
Comment (lunaticusgreen) [reply pmm-app/src/pmm-dbaas/components/DBCluster/EditDBClusterModal/EditDBClusterModal.tsx:92]: Seems like no, cleaned

File: pmm-app/src/pmm-dbaas/components/DBCluster/EditDBClusterModal/EditDBClusterModal.tsx (hunk 16 above)
Comment (lunaticusgreen) [reply pmm-app/src/pmm-dbaas/components/DBCluster/EditDBClusterModal/EditDBClusterModal.tsx:53]: yep, this is a workaround for TS, I don't know if there's a simpler way to do it

File: pmm-app/src/pmm-dbaas/components/DBCluster/EditDBClusterModal/EditDBClusterModal.tsx (hunk 7 above)
Comment (lunaticusgreen) [reply pmm-app/src/pmm-dbaas/components/DBCluster/EditDBClusterModal/EditDBClusterModal.tsx:59]: done, replaced with more specific type

**Review by tiagomotasantos — APPROVED** [review]
